File: data structures and algorithms/python/08_014.py
# Definition for singly-linked list

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

def reverse_list(head):
    prev = None
    current = head
    
    while current:
        # Save the next node
        next_node = current.next
        
        # Reverse the pointer
        current.next = prev
        
        # Move pointers forward
        prev = current
        current = next_node
    
    # prev is the new head
    return prev

# Helper function to create a linked list from an array
def create_linked_list(arr):
    if not arr:
        return None
    
    head = ListNode(arr[0])
    current = head
    
    for i in range(1, len(arr)):
        current.next = ListNode(arr[i])
        current = current.next
    
    return head

# Helper function to convert linked list to array
def linked_list_to_array(head):
    result = []
    current = head
    
    while current:
        result.append(current.val)
        current = current.next
    
    return result

# Example usage
list_head = create_linked_list([1, 2, 3, 4, 5])
reversed_head = reverse_list(list_head)
print(linked_list_to_array(reversed_head))  # [5, 4, 3, 2, 1]

File: data structures and algorithms/cs/08_014.cs
using System;

public class ReverseLinkedList
{
   // Definition for singly-linked list
   public class ListNode
   {
      public int val;
      public ListNode next;

      public ListNode(int x)
      {
         val = x;
      }
   }

   public static ListNode ReverseList(ListNode head)
   {
      ListNode prev = null;
      ListNode current = head;

      while (current != null)
      {
         // Save the next node
         ListNode next = current.next;

         // Reverse the pointer
         current.next = prev;

         // Move pointers forward
         prev = current;
         current = next;
      }

      // prev is the new head
      return prev;
   }

   // Helper function to create a linked list from an array
   public static ListNode CreateLinkedList(int[] arr)
   {
      if (arr == null || arr.Length == 0)
      {
         return null;
      }

      ListNode head = new ListNode(arr[0]);
      ListNode current = head;

      for (int i = 1; i < arr.Length; i++)
      {
         current.next = new ListNode(arr[i]);
         current = current.next;
      }

      return head;
   }

   // Helper function to convert linked list 
   // to array for display
   public static int[] LinkedListToArray(ListNode head)
   {
      // First, count the number of nodes
      int count = 0;
      ListNode current = head;
      while (current != null)
      {
         count++;
         current = current.next;
      }

      int[] result = new int[count];
      current = head;
      int i = 0;

      while (current != null)
      {
         result[i++] = current.val;
         current = current.next;
      }

      return result;
   }

   // Helper to print array
   public static void PrintArray(int[] arr)
   {
      Console.Write("[");
      for (int i = 0; i < arr.Length; i++)
      {
         Console.Write(arr[i]);
         if (i < arr.Length - 1)
         {
            Console.Write(", ");
         }
      }
      Console.WriteLine("]");
   }

   public static void Main(string[] args)
   {
      ListNode list = CreateLinkedList(
                         new int[] { 1, 2, 3, 4, 5 }
                      );
      ListNode reversedList = ReverseList(list);
      PrintArray(LinkedListToArray(reversedList));
      // [5, 4, 3, 2, 1]
   }
}